sdk/lib/drivers/wdf/wdfldr/class.c

```diff
@@ -0,0 +1,822 @@
+        if (NT_SUCCESS(status))
+        {
+            InitializeObjectAttributes(&objectAttributes, &versionString, OBJ_KERNEL_HANDLE | OBJ_CASE_INSENSITIVE, classVersionHandle, NULL);
+            status = ZwOpenKey(KeyHandle, KEY_QUERY_VALUE, &objectAttributes);
+            //*keyHandle = tmpKeyHandle;
+        }
+    }
+
+clean:
+    if (rootHandle)
+        ZwClose(rootHandle);
+    if (classVersionsHandle)
+        ZwClose(classVersionsHandle);
+    if (KeyHandle)
+        ZwClose(KeyHandle);
+
+    if (classVersions.Buffer)
+    {
+        ExFreePoolWithTag(classVersions.Buffer, WDFLDR_TAG);
+        classVersions.Length = 0;
+        classVersions.Buffer = NULL;
+    }
+
+    return status;
+}
+
+
+NTSTATUS
+NTAPI
+GetDefaultClassServiceName(
+    _In_ PWDF_CLASS_BIND_INFO ClassBindInfo,
+    _In_ PWDF_BIND_INFO BindInfo,
+    _Out_ PUNICODE_STRING ServiceName)
+{
+    PWCHAR pClassName;
+    PWCHAR pClassNameBegin;
+    WCHAR currentSymbol;
+    SIZE_T size;
+    NTSTATUS status;
+
+    pClassName = ClassBindInfo->ClassName;
+    pClassNameBegin = pClassName;
+    do
+    {
+        currentSymbol = *pClassName;
+        ++pClassName;
+    } while (currentSymbol);
+
+    size = sizeof(WCHAR) * (pClassName - pClassNameBegin - 1) + 148;
+    ServiceName->Buffer = ExAllocatePoolWithTag(PagedPool, size, WDFLDR_TAG);
+
+    if (ServiceName->Buffer == NULL)
+    {
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    ServiceName->MaximumLength = (USHORT)size;
+    status = RtlUnicodeStringPrintf(
+        ServiceName,
+        L"\\Registry\\Machine\\System\\CurrentControlSet\\Services\\%ws%02d%02d",
+        ClassBindInfo->ClassName,
+        ClassBindInfo->Version.Major,
+        BindInfo->Version.Major);
+
+    if (!NT_SUCCESS(status))
+    {
+        ExFreePoolWithTag(ServiceName->Buffer, WDFLDR_TAG);
+        ServiceName->Length = 0;
+        ServiceName->Buffer = NULL;
+    }
+
+    return status;
+}
+
+
+NTSTATUS
+NTAPI
+GetClassServicePath(
+    _In_ PWDF_CLASS_BIND_INFO ClassBindInfo,
+    _In_ PWDF_BIND_INFO BindInfo,
+    _Out_ PUNICODE_STRING ServicePath)
+{
+    NTSTATUS status;
+    HANDLE Handle = NULL;
+    PKEY_VALUE_PARTIAL_INFORMATION pKeyValPartial = NULL;
+    UNICODE_STRING ValueName = RTL_CONSTANT_STRING(L"Service");
+
+    ServicePath->Length = 0;
+    ServicePath->Buffer = NULL;
+
+    status = GetClassRegistryHandle(ClassBindInfo, BindInfo, &Handle);
+
+    if (!NT_SUCCESS(status))
+    {
+        __DBGPRINT(("ERROR: GetClassRegistryHandle failed with status 0x%x\n", status));        
+    }
+    else
+    {
+        status = FxLdrQueryData(Handle, &ValueName, WDFLDR_TAG, &pKeyValPartial);
+        
+        if (!NT_SUCCESS(status))
+        {
+            __DBGPRINT(("ERROR: QueryData failed with status 0x%x\n", status));
+        }
+        else
+        {
+            status = BuildServicePath(pKeyValPartial, 0x304C7846u, ServicePath);
```

Comment:
Please add some documentation or explain what's this magic construct for. Having this value as-is looks fishy.

Reply:
For information, this magic hexadecimal value corresponds to the "string" made DWORD: `0LxF` (aka `FxL0`)

Reply:
@HBelusca right. Here must be WDFLDR_TAG